libvmi/os/windows/core.c

```diff
@@ -661,6 +661,280 @@ get_kpgd_from_rekall_profile(vmi_instance_t vmi)
+     * */
+    selector >>= 3;
+
+    /* Now we need to determine address of the Global Descriptor Table (GDT)
+     * It is stored in GDTR_BASE register.
+     * */
+    addr_t gdtr = 0;
+    if (VMI_FAILURE == vmi_get_vcpureg(vmi, &gdtr, GDTR_BASE, 0))
+    {
+        return VMI_FAILURE;
+    }
+
+    gdtr += selector * 8;       // size of each entry in GDT is 8 bytes
+
+    /* Each entry has a complex structure, 
+     * that represents memory segment.
+     * We need to determine starting address
+     * of memory segment.
+     * 
+     * http://wiki.osdev.org/GDT
+     * */
+    uint8_t buffer[8] = {0};
+
+    bytes_readen = 0;
+    if (VMI_FAILURE == vmi_read_va(vmi, gdtr, 0, sizeof(buffer), buffer, &bytes_readen) ||
+        bytes_readen != sizeof(buffer))
+    {
+        return VMI_FAILURE;
+    }
+
+    uint32_t segment = (uint32_t)buffer[7] << 24
+                    | (uint32_t)buffer[4] << 16 
+                    | (uint32_t)buffer[3] << 8
+                    | buffer[2];
+
+    /* We have the starting address of the segment.
+    * Now we need to determine offset of Interrupt Service Routine
+    * inside of this segment, using IDT descriptor
+    * */
+
+    uint64_t address = 0;
+    page_mode_t mode = vmi_get_page_mode(vmi, 0);
+    switch(mode)
+    {
+        case VMI_PM_IA32E:
+            if (VMI_FAILURE == get_interrupt_routine_offset_64(vmi, idt_descriptor_address, &address))
+            {
+                return VMI_FAILURE;
+            }
+            break;
+        case VMI_PM_LEGACY:
+        case VMI_PM_PAE:
+            if (VMI_FAILURE == get_interrupt_routine_offset_32(vmi, idt_descriptor_address, &address)) 
+            {
+                return VMI_FAILURE;
+            }
+            break;
+        default:
+            return VMI_FAILURE;
+    };
+
+    *out_addr = address + segment;       // segment start + offset
+
+    return VMI_SUCCESS;
+}
+
+static bool
+nt_kernel_name_check(const char* name)
+{
+    return (0 == strncasecmp(name, "nt", 2) ||
+            0 == strncasecmp(name, "wrkx", 4) || 
+            0 == strncasecmp(name, "xnt", 3));
+}
+
+static status_t
+find_ntoskrnl_va(vmi_instance_t vmi, addr_t start_address_va, addr_t* ntoskrnl_address_va)
+{
+    status_t ret = VMI_FAILURE;
+    const size_t kernel_image_name_max_size = 8;
+    const uint64_t mask_4k = VMI_PS_4KB - 1;
+
+    uint8_t* buffer = malloc(VMI_PS_4KB);
+    if (NULL == buffer)
+    {
+        return ret;
+    }
+
+    struct export_table export_tbl;
+    addr_t et_rva = 0;
+    size_t et_size = 0;
+
+    access_context_t ctx =
+    {
+        .translate_mechanism = VMI_TM_PROCESS_PID,
+        .addr = 0,
+        .pid = 0
+    };
+
+    // virtual address of PE-image is always aligned to 4K boundaries
+    start_address_va &= ~mask_4k;
+
+    size_t bytes_readen = 0;
+    while (start_address_va > VMI_PS_4KB)       // do not check virtual address 0x1000, because there is no kernel there
```

Comment:
Going anywhere under `0xffff800000000000` is guaranteed to not have a kernel on 64-bit machines. On 32-bit going under `0x80000000` is also safe to assume to be the case (albeit that boundary being configurable).